src/main/java/com/google/googleidentity/oauth2/endpoint/TokenEndpoint.java

```diff
@@ -0,0 +1,139 @@
+/*
+    Copyright 2020 Google LLC
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+package com.google.googleidentity.oauth2.endpoint;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.googleidentity.oauth2.client.ClientDetailsService;
+import com.google.googleidentity.oauth2.exception.InvalidRequestException;
+import com.google.googleidentity.oauth2.exception.OAuth2Exception;
+import com.google.googleidentity.oauth2.exception.OAuth2ExceptionHandler;
+import com.google.googleidentity.oauth2.request.OAuth2Request;
+import com.google.googleidentity.oauth2.util.OAuth2ParameterNames;
+import com.google.googleidentity.oauth2.util.OAuth2Utils;
+import com.google.googleidentity.oauth2.validator.TokenEndpointRequestValidator;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.util.logging.Logger;
+
+/**
+ * Token exchange endpoint in OAuth2 Server
+ */
+@Singleton
+public class TokenEndpoint extends HttpServlet {
+
+    private static final long serialVersionUID = 5L;
+
+    private static final Logger log = Logger.getLogger("TokenEndpoint");
+
+    private final ClientDetailsService clientDetailsService;
+
+
+    @Inject
+    public TokenEndpoint(ClientDetailsService clientDetailsService) {
+        this.clientDetailsService = clientDetailsService;
+    }
+
+    protected void doGet(HttpServletRequest request, HttpServletResponse response)
+            throws ServletException, IOException, UnsupportedOperationException {
+
+        OAuth2Exception exception =
+                new InvalidRequestException(
+                        InvalidRequestException.ErrorCode.UNSUPPORTED_REQUEST_METHOD);
+        log.info("Token endpoint does not support GET request." );
+        response.setStatus(exception.getHttpCode());
+        response.setContentType("application/json;charset=UTF-8");
+        response.getWriter().println(
+                OAuth2ExceptionHandler.getResponseBody(exception).toJSONString());
+        response.getWriter().flush();
```

Comment:
Can you just throw InvalidRequestException and rely on OAuth2ExceptionHandler to format the response?

Same with doPost() validation exception below.